File: api/src/Models/EvenementSurvenu.php
<?php
declare(strict_types=1);

namespace App\Models;

use App\Database;

/**
 * EvenementSurvenu — event that occurred on public roads (Service Général).
 *
 * Single-table feature: the identities of the parties involved
 * (auteurs présumés / victimes / témoins) are stored as text fields
 * directly on the record.
 */
class EvenementSurvenu
{
    public static function all(array $filters = []): array
    {
        $db = Database::getInstance()->getConnection();
        $where = [];
        $args = [];

        if (!empty($filters['search'])) {
            $where[] = '(e.type_evenement LIKE ? OR e.lieu_exact LIKE ? OR e.auteurs_presumes LIKE ? OR e.victimes LIKE ? OR e.temoins LIKE ?)';
            $s = '%' . $filters['search'] . '%';
            array_push($args, $s, $s, $s, $s, $s);
        }
        if (!empty($filters['type'])) {
            $where[] = 'e.type_evenement = ?';
            $args[] = $filters['type'];
        }

        $sql = 'SELECT e.*, u.username AS agent_username, u.personnel_id AS agent_personnel_id, p.firstname AS agent_prenoms, p.lastname AS agent_nom
                FROM evenement_survenu e
                LEFT JOIN users u ON e.created_by = u.id
                LEFT JOIN personnel p ON u.personnel_id = p.id';
        if ($where) {
            $sql .= ' WHERE ' . implode(' AND ', $where);
        }
        $sql .= ' ORDER BY e.date_evenement DESC, e.id DESC';

        $stmt = $db->prepare($sql);
        $stmt->execute($args);
        return $stmt->fetchAll(\PDO::FETCH_ASSOC);
    }

    public static function find(int $id): ?array
    {
        $db = Database::getInstance()->getConnection();
        $stmt = $db->prepare(
            'SELECT e.*, u.username AS agent_username, u.personnel_id AS agent_personnel_id, p.firstname AS agent_prenoms, p.lastname AS agent_nom
             FROM evenement_survenu e
             LEFT JOIN users u ON e.created_by = u.id
             LEFT JOIN personnel p ON u.personnel_id = p.id
             WHERE e.id = ?'
        );
        $stmt->execute([$id]);
        $row = $stmt->fetch(\PDO::FETCH_ASSOC);
        return $row !== false ? $row : null;
    }

    public static function create(array $data): int
    {
        $db = Database::getInstance()->getConnection();
        $stmt = $db->prepare(
            'INSERT INTO evenement_survenu
                (date_evenement, heure_evenement, type_evenement, lieu_exact,
                 auteurs_presumes, victimes, temoins, mesures_prises, latitude, longitude, created_by)
             VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)'
        );
        $stmt->execute([
            $data['date_evenement'],
            $data['heure_evenement'],
            $data['type_evenement'],
            $data['lieu_exact'],
            $data['auteurs_presumes'] ?? null,
            $data['victimes'] ?? null,
            $data['temoins'] ?? null,
            $data['mesures_prises'] ?? null,
            $data['latitude'] ?? null,
            $data['longitude'] ?? null,
            $data['created_by'] ?? null,
        ]);
        return (int) $db->lastInsertId();
    }

    public static function update(int $id, array $data): bool
    {
        $db = Database::getInstance()->getConnection();
        $stmt = $db->prepare(
            'UPDATE evenement_survenu SET
                date_evenement = ?, heure_evenement = ?, type_evenement = ?, lieu_exact = ?,
                auteurs_presumes = ?, victimes = ?, temoins = ?, mesures_prises = ?,
                latitude = ?, longitude = ?
             WHERE id = ?'
        );
        return $stmt->execute([
            $data['date_evenement'],
            $data['heure_evenement'],
            $data['type_evenement'],
            $data['lieu_exact'],
            $data['auteurs_presumes'] ?? null,
            $data['victimes'] ?? null,
            $data['temoins'] ?? null,
            $data['mesures_prises'] ?? null,
            $data['latitude'] ?? null,
            $data['longitude'] ?? null,
            $id,
        ]);
    }

    public static function delete(int $id): bool
    {
        $db = Database::getInstance()->getConnection();
        $stmt = $db->prepare('DELETE FROM evenement_survenu WHERE id = ?');
        return $stmt->execute([$id]);
    }
}

File: android/app/src/main/java/com/gsoft/opus/domain/model/EvenementSurvenu.kt
package com.gsoft.opus.domain.model

/**
 * EvenementSurvenu — event that occurred on public roads (Service Général).
 *
 * Single-record feature: the identities of the parties involved
 * (auteurs présumés / victimes / témoins) are stored as text fields
 * directly on the record.
 */
data class EvenementSurvenu(
    val id: Int,
    val dateEvenement: String,
    val heureEvenement: String,
    val typeEvenement: String,
    val lieuExact: String,
    val auteursPresumes: String?,
    val victimes: String?,
    val temoins: String?,
    val mesuresPrises: String?,
    /** GPS latitude captured at record time (mobile only, null on desktop). */
    val latitude: Double?,
    /** GPS longitude captured at record time (mobile only, null on desktop). */
    val longitude: Double?,
    val createdBy: Int?,
    val createdAt: String?,
    val updatedAt: String?,
    val agentUsername: String?,
    val agentPrenoms: String?,
    val agentNom: String?
) {
    /** "HH:MM:SS" from the API is displayed as "HH:MM". */
    val heureDisplay: String get() = heureEvenement.take(5)

    /**
     * The type is stored as the catalog label verbatim. TYPE_LABELS only
     * maps legacy machine codes for rows created before the catalog.
     */
    val typeLabel: String get() = TYPE_LABELS[typeEvenement] ?: typeEvenement

    /** Display name of the agent who recorded the entry. */
    val agentDisplayName: String
        get() = listOfNotNull(agentPrenoms, agentNom)
            .filter { it.isNotBlank() }
            .joinToString(" ")
            .ifBlank { agentUsername ?: "" }

    companion object {
        const val TYPE_INFRACTION = "infraction"
        const val TYPE_INCIDENT = "incident"
        const val TYPE_ACCIDENT = "accident"
        const val TYPE_AUTRE = "autre"

        /** Ordered type codes for dropdowns. */
        val TYPES = listOf(TYPE_INFRACTION, TYPE_INCIDENT, TYPE_ACCIDENT, TYPE_AUTRE)

        /** French labels keyed by type code. */
        val TYPE_LABELS = mapOf(
            TYPE_INFRACTION to "Infraction",
            TYPE_INCIDENT to "Incident",
            TYPE_ACCIDENT to "Accident",
            TYPE_AUTRE to "Autre"
        )
    }
}

/** Attached file of an évènement survenu (stored on disk server-side). */
data class EvenementSurvenuAttachment(
    val id: Int,
    val evenementId: Int,
    val title: String,
    val filename: String,
    val originalFilename: String,
    val mimeType: String?,
    val fileSize: Long?,
    val createdAt: String?
)
